### ISSUE TITLE
New generic constraint error in 2.16.1

### ISSUE THREAD

**Issue by Alxandr** [fscheck/FsCheck#587]
I'm not sure what to do with this, so I figured reporting here and see if anyone knew would be my best course of action. We have a bot (renovate) that updates dependencies of projects at work, and the update from 2.16.0 to 2.16.1 of FsCheck broke our tests with the following error:

```
System.TypeLoadException : GenericArguments[0], 'b', on 'FsCheck.ReflectArbitrary+shrinkChildren@183-1[a]' violates the constraint of type parameter 'a'.
Stack Trace:
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Object[] arguments, Signature sig, Boolean constructor)
   at System.Reflection.RuntimeMethodInfo.UnsafeInvokeInternal(Object obj, Object[] parameters, Object[] arguments)
   at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
   at System.Reflection.MethodBase.Invoke(Object obj, Object[] parameters)
   at FsCheck.Runner.checkMethod(Config config, MethodInfo m, FSharpOption`1 target)
   at FsCheck.Check.Method(Config config, MethodInfo methodInfo, FSharpOption`1 target)
   at <StartupCode$FsCheck-Xunit>.$PropertyAttribute.testExec@199(PropertyTestCase this, IMessageBus messageBus, Object[] constructorArguments, CancellationTokenSource cancellationTokenSource, XunitTest test, RunSummary summary, TestOutputHelper outputHelper, Unit unitVar0)
----- Inner Stack Trace -----
   at FsCheck.ReflectArbitrary.reflectShrinkObj@226-7.Invoke(FSharpFunc`2 arg00)
   at Microsoft.FSharp.Core.FSharpFunc`2.InvokeFast[V,W,X](FSharpFunc`2 func, T arg1, TResult arg2, V arg3, W arg4)
   at FsCheck.ReflectArbitrary.reflectShrink[a,a](FSharpFunc`2 getShrink, a a)
   at FsCheck.Arb.Derive@1326.Shrinker(a a)
   at FsCheck.Testable.props@143-1[a,b,c](FSharpFunc`2 shrink, FSharpFunc`2 pf, a x)
   at FsCheck.Testable.shrinking[a,b,c](FSharpFunc`2 shrink, a x, FSharpFunc`2 pf)
   at FsCheck.Testable.forAll@163-1.Invoke(a _arg1)
   at FsCheck.GenBuilder.Bind@88.Invoke(Int32 n, StdGen r0)
   at FsCheck.Runner.test@96-1.GenerateNext(IEnumerable`1& next)
   at Microsoft.FSharp.Core.CompilerServices.GeneratedSequenceBase`1.MoveNextImpl()
   at Microsoft.FSharp.Core.CompilerServices.GeneratedSequenceBase`1.System-Collections-IEnumerator-MoveNext()
   at FsCheck.Common.loop@53-1.GenerateNext(IEnumerable`1& next)
   at Microsoft.FSharp.Core.CompilerServices.GeneratedSequenceBase`1.MoveNextImpl()
   at Microsoft.FSharp.Core.CompilerServices.GeneratedSequenceBase`1.System-Collections-IEnumerator-MoveNext()
   at FsCheck.Common.takeWhilePlusLast@61.GenerateNext(IEnumerable`1& next)
   at Microsoft.FSharp.Core.CompilerServices.GeneratedSequenceBase`1.MoveNextImpl()
   at Microsoft.FSharp.Core.CompilerServices.GeneratedSequenceBase`1.System-Collections-IEnumerator-MoveNext()
   at Microsoft.FSharp.Collections.SeqModule.Fold[T,TState](FSharpFunc`2 folder, TState state, IEnumerable`1 source)
   at FsCheck.Runner.runner[a](Config config, a prop)
   at FsCheck.Runner.check[a](Config config, a p)
```

**Comment by kurtschelfthout** [comment]
Strange - the only change between 2.16.0 and 2.16.1 is this: https://github.com/fscheck/FsCheck/commit/2616ed3b7ec7b752520d1a88164cf8489689f021 - not immediately obvious if or how that is related.

Are you by any chance calling `Arb.register` or passing types into `PropertyAttribute` that inherit? A situation like https://github.com/fscheck/FsCheck/pull/584

If you could share some context or a repro that would help too.

**Comment by Alxandr** [comment]
I'll look into it a bit more shortly. We definitely use `Arb.Register` (C#), and also use `PropertyAttribute`. I'll hopefully get back with more details within a few hours.

**Comment by Choc13** [comment]
I'm also hitting this when updating from `2.16.0` to `2.16.1`.

It seems to be failing whenever a custom type is used as an input to a test case, which previously FsCheck would have just automatically created instances for. Here's a minimal example:

```fsharp
type MyRecord = { X: bool }

[<Property>]
let testingNewFsCheck (x: MyRecord) = test <@ x = x @>
```

If we change the input to `(x: string)` then it works fine. I wonder if the change in 2616ed3 has introduced a regression that means things don't work correctly for FSharp types like Records and DUs? It seems to work fine with custom classes, e.g. this passes:

```fsharp
type MyClass(foo: string) =
    member _.Foo = foo

[<Property>]
let testingNewFsCheck (x: MyClass) = test <@ x = x @>
```

**Comment by PierreYvesR** [comment]
Same here:
Note that a single `float[]` parameter is fine. It breaks when another parameter is added.
```f#
[<Property>]
let myProperty(s:float, v:float[]) = true

```
[Full test case here.](https://github.com/PierreYvesR/FsCheckBug)

**Comment by Choc13** [comment]
I wonder if that's because in that case you've tupled the args and actually this change has also regressed the behaviour for tuple params too.

**Comment by kurtschelfthout** [comment]
It's strange, I can't reproduce on master but I can reproduce with the build from NuGet. Thanks for the repro I'll investigate further.

**Comment by kurtschelfthout** [comment]
Given it's definitely properly broken I'll also unlist the 2.16.1 build from NuGet while investigation is underway.

**Comment by kurtschelfthout** [comment]
Fix released in 2.16.2

**Comment by Alxandr** [comment]
Great. Sorry I couldn't get back to you, but good that this has been fixed :).
